core/src/Streamly/Internal/Data/Array/Type.hs

```diff
@@ -324,6 +324,19 @@ fromStreamD str = unsafeFreeze <$> MA.fromStreamD str
 -- Cloning
 -------------------------------------------------------------------------------
 
+-- | @clone@ creates new immutable array from a given slice. The size of the
+-- underlying raw array will be equal to the size of the given slice.
+--
+-- >>> import Streamly.Internal.Data.MutByteArray as MBA
+-- >>> import Streamly.Internal.Data.Array as Array
+-- >>> array = Array.fromList ['a', 'b', 'c', 'd', 'e', 'f']
+-- >>> slice1 = Array.getSliceUnsafe 1 4 array
+-- >>> slice2 = Array.clone slice1
+-- >>> MBA.sizeOfMutableByteArray (Array.arrContents slice1)
+-- 6
+-- >>> MBA.sizeOfMutableByteArray (Array.arrContents slice2)
+-- 4
+--
```

Comment:
We can pin or unpin a slice rather than cloning it.

Reply:
Currently, `pin` on a pinned array is noop and `unpin` on an unpinned array is a noop.
`pin` on an unpinned array == `pinnedClone`
`unpin` on a pinned array == `clone`

Do we want `pin` and `unpin` to allocate array everytime?

Reply:
The original question was why would you want to clone an immutable array?